File: CZECHMATE/CZECHMATE/Features/Game/Components/GameEndReportTiming.swift
//
//  GameEndReportTiming.swift
//  CZECHMATE — Časové řady pro souhrn partie (grafy z razítek tahů).
//

import Foundation

/// Jedna „polotahová“ hodnota: čas od předchozího tahu podle `timestamp` z desky.
struct EndgameThinkPlyPoint: Identifiable, Equatable {
    var id: Int { plyIndex }
    /// 1 = první tah bílého …
    let plyIndex: Int
    let isWhite: Bool
    /// Sekundy od předchozího záznamu v historii; u prvního tahu `nil`, pokud není čas před startem.
    let secondsFromPrevious: Double?
}

/// Bod křivky „kolik minut uběhlo od začátku“ po daném půltahu.
struct EndgameCumulativePoint: Identifiable, Equatable {
    var id: Int { plyIndex }
    let plyIndex: Int
    let totalSeconds: Double
}

enum GameEndReportTiming {
    /// Odvozeno z `timestamp` u tahů (stejné jednotky jako firmware — obvykle sekundy od startu MCU / partie).
    static func thinkPlyPoints(from moves: [GameHistoryMove]) -> [EndgameThinkPlyPoint] {
        moves.enumerated().map { i, m in
            let prev: Double?
            if i == 0 {
                prev = nil
            } else {
                let a = moves[i - 1].timestamp
                let b = m.timestamp
                if let t0 = a, let t1 = b, t1 >= t0 {
                    prev = Double(t1 - t0)
                } else {
                    prev = nil
                }
            }
            return EndgameThinkPlyPoint(
                plyIndex: i + 1,
                isWhite: i % 2 == 0,
                secondsFromPrevious: prev
            )
        }
    }

    static func cumulativePoints(from think: [EndgameThinkPlyPoint]) -> [EndgameCumulativePoint] {
        var sum: Double = 0
        var out: [EndgameCumulativePoint] = []
        for p in think {
            if let d = p.secondsFromPrevious {
                sum += d
                out.append(EndgameCumulativePoint(plyIndex: p.plyIndex, totalSeconds: sum))
            }
        }
        return out
    }

    static func shareTimingLines(from moves: [GameHistoryMove], maxLines: Int = 24) -> [String] {
        let pts = thinkPlyPoints(from: moves)
        guard pts.contains(where: { $0.secondsFromPrevious != nil }) else { return [] }
        var lines: [String] = ["Čas mezi tahy (z razítek desky):"]
        for p in pts.prefix(maxLines) {
            let side = p.isWhite ? "B" : "Č"
            if let s = p.secondsFromPrevious {
                lines.append("  \(p.plyIndex).\(side)  \(formatSeconds(s))")
            } else {
                lines.append("  \(p.plyIndex).\(side)  —")
            }
        }
        if pts.count > maxLines {
            lines.append("  … (+ \(pts.count - maxLines) tahů)")
        }
        return lines
    }

    static func formatSeconds(_ s: Double) -> String {
        if s < 60 { return String(format: "%.1f s", s) }
        let m = Int(s) / 60
        let r = s - Double(m * 60)
        if r < 0.5 { return "\(m) min" }
        return String(format: "%d min %.0f s", m, r)
    }

    /// Průměrný čas na tah (sekundy) podle strany — jen tahy s `secondsFromPrevious`.
    static func averageSecondsPerSide(from think: [EndgameThinkPlyPoint]) -> (white: Double?, black: Double?) {
        var wSum = 0.0, wCount = 0
        var bSum = 0.0, bCount = 0
        for p in think {
            guard let s = p.secondsFromPrevious else { continue }
            if p.isWhite {
                wSum += s
                wCount += 1
            } else {
                bSum += s
                bCount += 1
            }
        }
        let wAvg = wCount > 0 ? wSum / Double(wCount) : nil
        let bAvg = bCount > 0 ? bSum / Double(bCount) : nil
        return (wAvg, bAvg)
    }

    /// Celkový průměr přes všechny měřené půltahy.
    static func overallAverageSeconds(from think: [EndgameThinkPlyPoint]) -> Double? {
        let vals = think.compactMap(\.secondsFromPrevious)
        guard !vals.isEmpty else { return nil }
        return vals.reduce(0, +) / Double(vals.count)
    }
}

File: flutter_czechmate/lib/core/utils/game_end_report_timing.dart
import '../models/game_snapshot.dart';

/// Parita `GameEndReportTiming.swift` / `EndgameThinkPlyPoint`.
class GameEndThinkPlyPoint {
  const GameEndThinkPlyPoint({
    required this.plyIndex,
    required this.isWhite,
    this.secondsFromPrevious,
  });

  final int plyIndex;
  final bool isWhite;
  final double? secondsFromPrevious;
}

class GameEndCumulativePoint {
  const GameEndCumulativePoint({required this.plyIndex, required this.totalSeconds});

  final int plyIndex;
  final double totalSeconds;
}

class GameEndReportTiming {
  /// Oříznutí nesmyslných intervalů (chyba hodin / reboot desky) — graf bez vertikálních artifactů.
  static const double maxPlausibleMoveSeconds = 3600;

  static double? _sanitizeMoveDeltaSec(double raw) {
    if (raw.isNaN || raw.isInfinite || raw < 0) return null;
    if (raw > maxPlausibleMoveSeconds) return null;
    return raw;
  }

  static List<GameEndThinkPlyPoint> thinkPlyPoints(List<GameHistoryMove> moves) {
    final out = <GameEndThinkPlyPoint>[];
    for (var i = 0; i < moves.length; i++) {
      final m = moves[i];
      double? prev;
      if (i == 0) {
        prev = null;
      } else {
        final a = moves[i - 1].timestamp;
        final b = m.timestamp;
        if (a != null && b != null && b >= a) {
          prev = _sanitizeMoveDeltaSec((b - a) / 1000.0);
        } else {
          prev = null;
        }
      }
      out.add(GameEndThinkPlyPoint(plyIndex: i + 1, isWhite: i % 2 == 0, secondsFromPrevious: prev));
    }
    return out;
  }

  static List<GameEndCumulativePoint> cumulativePoints(List<GameEndThinkPlyPoint> think) {
    var sum = 0.0;
    final out = <GameEndCumulativePoint>[];
    for (final p in think) {
      final d = p.secondsFromPrevious;
      if (d != null) {
        sum += d;
        out.add(GameEndCumulativePoint(plyIndex: p.plyIndex, totalSeconds: sum));
      }
    }
    return out;
  }
}
